Mono.Debugger.Soft/Mono.Debugger.Soft/VirtualMachine.cs

```diff
@@ -177,30 +177,32 @@ internal void AddThreadToInvalidateList (ThreadMirror threadMirror)
 		}
 
 		public IList<ThreadMirror> GetThreads () {
-			var threads = threadCache;
-			if (threads == null) {
-				long[] ids = null;
-				var fetchingEvent = new ManualResetEvent (false);
-				vm.conn.VM_GetThreads ((threadsIds) => {
-					ids = threadsIds;
-					threadCache = threads = new ThreadMirror [threadsIds.Length];
-					fetchingEvent.Set ();
-				});
-				if (WaitHandle.WaitAny (new []{ vm.conn.DisconnectedEvent, fetchingEvent }) == 0) {
-					throw new VMDisconnectedException ();
+			lock (threadCacheLocker) {
+				var threads = threadCache;
+				if (threads == null) {
+					long[] ids = null;
+					var fetchingEvent = new ManualResetEvent (false);
+					vm.conn.VM_GetThreads ((threadsIds) => {
+						ids = threadsIds;
+						threadCache = threads = new ThreadMirror [threadsIds.Length];
```

Comment:
Maybe this can be fixed without locking by chaning `threadCache = threads = new ThreadMirror [threadsIds.Length];` to `threads = new ThreadMirror [threadsIds.Length];`